lib/Sema/CSDiag.cpp

```diff
@@ -1030,15 +1024,16 @@ static bool findGenericSubstitutions(DeclContext *dc, Type paramType, Type actua
       if (dc && original->isTypeParameter())
         original = ArchetypeBuilder::mapTypeIntoContext(dc, original);
       
-      if (auto archetype = original->getAs<ArchetypeType>()) {
-        archetypes.push_back(archetype);
-        substitutions.push_back(substitution->getReplacementType());
-      }
+      Type replacement = substitution->getReplacementType();
+      if (auto archetype = original->getAs<ArchetypeType>())
+        archetypesMap[archetype] = replacement;
+      if (auto reverse = replacement->getAs<ArchetypeType>())
+        archetypesMap[reverse] = original;
```

Comment:
I'm struggling to come up with a concise way to put it into words so I can add a comment there. Given code like this:
```
protocol EqualComparable {
  func isEqual(other: Self) -> Bool
}

func doCompare<T : EqualComparable, U : EqualComparable>(t: T, u: U) -> Bool {
  return t1.isEqual(u) // expected-error {{cannot invoke 'isEqual' with an argument list of type '(U)'}} expected-note {{expected an argument list of type '(T)'}}
}
```
The paramType that gets passed into here is a substitution with original=EqualComparable.Self and replacement=T. Then in GenericVisitor, the mismatch is between paramType=T (because the TypeMatcher canonicalizes and so sees the replacement type) and actual argType=U. 

So I need to capture both sides of the equivalency between Self and T here in order to find the mismatch.

Any suggestions on how to express that? Am I making sense at all? :-)